connectors/stash/src/main/java/com/vmware/connectors/stash/StashController.java

```diff
@@ -0,0 +1,444 @@
+        headers.add(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE);
+
+        final String url = String.format(StashConstants.STASH_PR_COMMENT_URL_FORMAT,
+                baseUrl,
+                pullRequest.getProjectKey(),
+                pullRequest.getRepositorySlug(),
+                pullRequest.getPullRequestId());
+
+        final StashComment stashComment = new StashComment(comment);
+
+        final ListenableFuture<ResponseEntity<String>> response = this.rest.exchange(
+                url,
+                HttpMethod.POST,
+                new HttpEntity<>(stashComment, headers),
+                String.class);
+        return Async.toSingle(response);
+    }
+
+    private List<String> getComments(final String baseUrl,
+                                     final HttpHeaders headers,
+                                     final StashPullRequest pullRequest) {
+        final String url = String.format(StashConstants.STASH_ACTIVITIES_URL_FORMAT,
+                baseUrl,
+                pullRequest.getProjectKey(),
+                pullRequest.getRepositorySlug(),
+                pullRequest.getPullRequestId());
+
+        final ListenableFuture<ResponseEntity<JsonDocument>> response = this.rest.exchange(
+                url,
+                HttpMethod.GET,
+                new HttpEntity<>(headers),
+                JsonDocument.class);
+
+        final JsonDocument[] jsonDocuments = new JsonDocument[1];
+        Async.toSingle(response)
+                .subscribe(entity -> {
+                    jsonDocuments[0] = entity.getBody();
+                });
+        Assert.isTrue(Objects.nonNull(jsonDocuments[0]), "JsonDocument should not be null.");
+
+        final List<String> comments = jsonDocuments[0].read("$.values[*].comment.text");
+        return comments.stream()
+                .limit(COMMENTS_SIZE)
+                .collect(Collectors.toList());
+    }
+
+    private Single<ResponseEntity<String>> performStashAction(final String baseUrl,
+                                                              final String authHeader,
+                                                              final StashPullRequest pullRequest,
+                                                              final StashAction stashAction) {
+        final HttpHeaders headers = buildHeaders(authHeader);
+
+        // Get current version of the pull request. Pull request "version" changes when we do any actions on it.
+        // When the pull request is raised, the current value will be 0.
+        // For example, when we approve the pull request, then the version will change from 0 to 1.
+        // We have to add the latest version of the pull request URI to do any actions. Otherwise, the action will be rejected.
+        // If the build for the branch is going on, then the actions would be rejected.
+        final String version = getVersion(headers, baseUrl, pullRequest);
```

Comment:
Do you know if this is for all actions or just some?

In github it was only for merge (optional), in gitlab it was for approve and merge (optional for both).

I remember bitbucket being really obnoxious about some operations in its UI when someone changed something while you were reviewing, but not others.

Also, in github & gitlab, I chose to look up the version when making the cards and to pass that version (commit hash actually) via request-params in the cards' action.  This would cause clicking a button on an old card to error instead of merge something that has changed.

We probably want to do the same for this (or maybe don't specify the hashes in github/gitlab).

It's not clear to me how often the client will ask for an updated card, so maybe the details of this are a bigger discussion for a later time (and including @davidpshaw @jonathanrfisher1).

Reply:
The version changes for all the actions in Bitbucket.. To be on safer side, version is fetched everytime before executing an action..

Reply:
Yuck!  If they change the version on comments, approves, and rejects then yea we have to look it up before every action (the client isn't going to ask for a new card in-between the user commenting and then approving).

It sounds like Atlassian is a little too aggressive on their version checking :)